### ISSUE TITLE
Presence of `django-stubs` breaks type checking of `djangorestframework-dataclasses`

### ISSUE THREAD

**Issue by johnthagen** [typeddjango/django-stubs#1810]
# Bug report

## What's wrong

With this example code:

```py
from dataclasses import dataclass

from rest_framework_dataclasses.serializers import DataclassSerializer


@dataclass
class Person:
    name: str
    age: int


class PersonSerializer(DataclassSerializer[Person]):
    class Meta:
        dataclass = Person


serializer = PersonSerializer(data={"name": "Tony", "age": 39})
serializer.is_valid()
person = serializer.validated_data
print(f"{person.name}")
```

With the following Mypy configuration:

```toml
[tool.mypy]
strict = true
```

Mypy produces a false positive error:

```
(venv) $ mypy main.py
main.py:20: error: "T" has no attribute "name"  [attr-defined]
```

when [`django-stubs`](https://github.com/typeddjango/django-stubs) is installed. I haven't been able to piece together _why_ this is, or if this is a bug with Mypy itself?

Does `django-stubs` do anything in particular that could trigger this behaviour?

### Environment

- Python 3.10.11

```
Package                         Version
------------------------------- ------------
asgiref                         3.7.2
Django                          4.2.6
django-stubs                    4.2.6
django-stubs-ext                4.2.5
djangorestframework             3.14.0
djangorestframework-dataclasses 1.3.1
mypy                            1.6.1
mypy-extensions                 1.0.0
pip                             23.1.2
pytz                            2023.3.post1
setuptools                      68.1.2
sqlparse                        0.4.4
tomli                           2.0.1
types-pytz                      2023.3.1.1
types-PyYAML                    6.0.12.12
typing_extensions               4.8.0
wheel                           0.40.0
```

These combinations also trigger the error:

- Mypy 1.6.1, django-stubs 4.2.5
- Mypy 1.5.1, django-stubs 4.2.4
- Mypy 1.4.1, django-stubs 4.2.3
- Mypy 1.6.1, django-stubs 4.2.2
- Mypy 1.3.0, django-stubs 4.2.1
- Mypy 1.2.0, django-stubs 4.2.0

## How is that should be

No type checking error should be raised. If `django-stubs` is not installed, inference within Mypy works properly.

## System information

- OS: macOS 13
- `python` version: 3.10
- `django` version: N/A
- `mypy` version: 1.6.1
- `django-stubs` version: 4.2.6
- `django-stubs-ext` version: 4.2.5

## Context

Previously reported and discussed at

- https://github.com/oxan/djangorestframework-dataclasses/issues/91

**Comment by johnthagen** [comment]
If this should be a Mypy bug report instead, please let me know. I thought I'd start here first.

**Comment by flaeppe** [comment]
Could this be related to recent work in `django-stubs` regarding `from django.utils.functional import cached_property`? Or is that too new in regards to the mentioned versions?

CC @intgr @UnknownPlatypus

Given that it's used in an example in the linked PR: https://github.com/oxan/djangorestframework-dataclasses/issues/91#issuecomment-1787752276

Otherwise I'm not seeing how this would or could be related to `django-stubs`, at least not with the information found in linked issues.

Example inlined for clarity:

> ```python
> from __future__ import annotations
>
> from typing import Any, Generic, TypeVar
> from django.utils.functional import cached_property
> import rest_framework.serializers
>
> T = TypeVar('T')
>
> class DataclassSerializer(rest_framework.serializers.Serializer, Generic[T]):
>     def __init__(self, *args: Any, **kwargs: Any):
>         pass
>
>     @cached_property
>     def validated_data(self) -> T:
>         pass
> ```

(Maybe unrelated, but `__init__` is missing `-> None:` in this example to be a fully typed signature, but `strict = true` is mentioned as config..)

I'm also seeing code related to `rest_framework` but I'm not seeing any mention of an installation of `djangorestframework-stubs`. Is or could that be a problem? (e.g. untyped objects)

**Comment by UnknownPlatypus** [comment]
older django-stubs seem to also trigger this issue so I'm not sure it's related 🤔 
> These combinations also trigger the error:
> Mypy 1.5.1, django-stubs 4.2.4
> Mypy 1.4.1, django-stubs 4.2.3

**Comment by intgr** [comment]
TL;DR: You're missing `djangorestframework-stubs`

1. Without `django-stubs` installed, mypy doesn't know the `cached_property` decorator, so `DataclassSerializer.validated_data` becomes `Any`, thus no error.
2. With `django-stubs` installed, `DataclassSerializer.validated_data` is now typed as property `T`, but looks like it's having problems resolving the `T` TypeVar.
3. Solution: Install `djangorestframework-stubs` as well, mypy can resolve `T` and everything works again.

I'm not sure why the resolution of T fails in (2). But  `DataclassSerializer` derives from `rest_framework.serializers.Serializer` which is untyped if you don't have DRF-stubs installed. It's not entirely surprising that a typed module which relies on another untyped module causes issues.

**Comment by flaeppe** [comment]
Nice hunt down. I would very much suspect that e.g. mypy states undefined behaviour for declaring generic classes subclassing untyped/unknown objects.

**Comment by johnthagen** [comment]
@intgr Thanks for tracking this down. I can confirm that installing `djangorestframework-stubs (3.14.4)` fixes the problem.

I will work with `djangorestframework-dataclasses` to document this typing dependency.
